Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/FilterAdapter.cpp

```diff
@@ -0,0 +1,501 @@
+                                cullUnmatchedChildRows(*valueIter, *matchIter);
+                            }
+                            else
+                            {
+                                // neither the node nor its children match, cull the value from the returned tree
+                                valueIter = rowValue.ArrayErase(valueIter);
+                                continue; // we've already moved valueIter forward, skip the ++valueIter below
+                            }
+                        }
+                        ++valueIter;
+                    }
+                }
+            };
+            // filter downwards from root
+            cullUnmatchedChildRows(filteredContents, m_root);
+        }
+        return filteredContents;
+    }
+
+    void RowFilterAdapter::HandleReset()
+    {
+        delete m_root;
+        m_root = NewMatchInfoNode(nullptr);
+        auto sourceContents = m_sourceAdapter->GetContents();
+
+        // we can assume all direct children of an adapter must be rows; populate each of them
+        for (size_t topIndex = 0; topIndex < sourceContents.ArraySize(); ++topIndex)
+        {
+            PopulateNodesAtPath(Dom::Path({Dom::PathEntry(topIndex)}));
+        }
+        NotifyResetDocument();
+    }
+
+    void RowFilterAdapter::HandleDomChange(const Dom::Patch& patch)
+    {
+        auto sourceContents = m_sourceAdapter->GetContents();
+        for (auto operationIterator = patch.begin(), endIterator = patch.end(); operationIterator != endIterator; ++operationIterator)
+        {
+            const auto& patchPath = operationIterator->GetDestinationPath();
+            auto rowPath = GetRowPath(patchPath);
+
+            if (operationIterator->GetType() == AZ::Dom::PatchOperation::Type::Remove)
+            {
+                auto matchingRow = GetMatchNodeAtPath(rowPath);
+                if (rowPath == patchPath)
+                {
+                    // node being removed is a row. We need to remove it from our graph
+                    auto& parentContainer = matchingRow->m_parentNode->m_childMatchState;
+                    parentContainer.erase(parentContainer.begin() + (patchPath.end() - 1)->GetIndex());
+
+                    // update former parent's match state, as its m_matchingDescendents may have changed
+                    UpdateMatchState(matchingRow->m_parentNode);
+                }
+                else
+                {
+                    // removed node wasn't a row, so we need to re-cache the owning row's cached info
+                    CacheDomInfoForNode(sourceContents[rowPath], matchingRow);
+                    UpdateMatchState(matchingRow);
+                }
+            }
+            else if (operationIterator->GetType() == AZ::Dom::PatchOperation::Type::Replace)
+            {
+                // replace operations can change child values from or into rows. Handle all cases
```

Comment:
There's... an argument that JSON-patch was too permissive a delta comms structure, but this is well handled.

Reply:
ah yes, an argument from a bygone era. We'll just have to keep going with what we've got...